open-apparel-registry/open-apparel-registry
# Add update facility location dashboard page

## Overview

Adds a new API and dashboard page to update the coordinates of a `Facility`.

Connects #738

## Demo

![2019-09-18 19 50 09](https://user-images.githubusercontent.com/17363/65209728-c0977880-da4d-11e9-91e5-b1ac08edbd46.gif)

## Notes

Changes to the facility location made by superusers will be recorded as records in the new `FacilityLocation` table, allowing us to capture details about the moderation activity.

We exclude the new endpoint from the Swagger docs because it is intended to be consumed by the superuser-only dashboard.

The layout of the new dashboard page is based on other "two panel" pages, like facility merging.

We added "Location" to the `DashboardFacilityCard` so the existing coordinates are visible. This adds the location to every dashboard page that uses the `DashboardFacilityCard` but showing extra information is not a problem.

## Testing Instructions

* Search the facilities and copy an OAR ID.
* Browse http://localhost:6543/dashboard/updatefacilitylocation.
* Enter the OAR ID and click search. Verify that the facility details are loaded.
* Copy the longitude and latitude from the facility details to the "Longitude" and "Latitude" fields and verify that a map showing the same location is loaded. Ensure that there are only 5 decimal places.
* Adjust the coordinates to move the point.
* Click "Update Location." Verify that the map updates, the New Location fields are cleared, and a success toast message is shown.
* Repeat the process, this time filling in the "Contributor" and "Notes" fields. Click "Update Location."
* Run `./scripts/dbshell` and run `SELECT * FROM api_facilitylocation;`. Verify that the table contains rows for each of the saved location changes.

## Checklist
